# Dealing with missing data with Python Pandas library

Many datasets will have some amount of missing data in real world examples. 

In this summary note, we will discuss some general considerations for missing data and demonstrate some built-in Pandas tools for handling missing data in Python.


In Pandas missing data is represented by two value:

  None: is a Python singleton object that is often used for missing data in Python code.

  NaN : NaN (an acronym for Not a Number), is a special floating-point value recognized by all systems that use the standard     IEEE floating-point representation



Pandas treat None and NaN as essentially interchangeable for indicating missing or null values. To facilitate this convention, there are several useful functions for detecting, removing, and replacing null values in Pandas DataFrame:

isnull(): Generate a boolean mask indicating missing values

notnull(): Opposite of isnull()

dropna(): Return a filtered version of the data

fillna(): Return a copy of the data with missing values filled or imputed

replace(): Replace a string, regex, list, dictionary, series, number etc. from a dataframe.This is a very rich function 
          as it has many variations.

interpolate(): Fill NA values in the dataframe or series. This is a very powerful function to fill the missing values.
            It uses various interpolation technique to fill the missing values. 
            
            
SPECIAL NOTE: This note is a completion of various website resources regarding the subject. It tries to provide you with a good summary of the generally known basic functions and how they are used with examples.The resources are cited at the end. Scientific referencing rules were not followed.  

# Checking for missing values using isnull() and notnull()

We know that Pandas fill in the blank space with “NA”. Using the isnull() method, we can confirm that both the missing value and “NA” were recognized as missing values (np.NaN (NumPy Not a Number) and the Python None). 

Pandas will recognize both empty cells and “NA” types as missing values. 

In [8]:
import pandas as pd  

data = pd.read_csv("employees.csv")  

# creating bool series True for NaN values  
bool_series = pd.isnull(data["Gender"])  

# displaying data only with Gender = NaN  
data[bool_series].head()  


First Name Gender Start Date Last Login Time  Salary  Bonus %  \
20       Lois    NaN  4/22/1995         7:18 PM   64714    4.934   
22     Joshua    NaN   3/8/2012         1:58 AM   90816   18.816   
27      Scott    NaN  7/11/1991         6:58 PM  122367    5.218   
31      Joyce    NaN  2/20/2005         2:40 PM   88657   12.752   
41  Christine    NaN  6/28/2015         1:08 AM   66582   11.308   

   Senior Management                  Team  
20              True                 Legal  
22              True       Client Services  
27             False                 Legal  
31             False               Product  
41              True  Business Development

Sometimes it might be the case where there’s missing values that have different formats. For example: 'n/a', 'NA','—', 'na' 

Let us change the data below and execute the same codes above:

In [9]:
data.iloc[20,1]='n/a'

In [10]:
data[20:22]

First Name Gender Start Date Last Login Time  Salary  Bonus %  \
20       Lois    n/a  4/22/1995         7:18 PM   64714    4.934   
21    Matthew   Male   9/5/1995         2:12 AM  100612   13.645   

   Senior Management       Team  
20              True      Legal  
21             False  Marketing

In [11]:
# creating bool series True for NaN values  
bool_series = pd.isnull(data["Gender"])  
    
# displaying data only with Gender = NaN  
data[bool_series].head()

First Name Gender Start Date Last Login Time  Salary  Bonus %  \
22     Joshua    NaN   3/8/2012         1:58 AM   90816   18.816   
27      Scott    NaN  7/11/1991         6:58 PM  122367    5.218   
31      Joyce    NaN  2/20/2005         2:40 PM   88657   12.752   
41  Christine    NaN  6/28/2015         1:08 AM   66582   11.308   
49      Chris    NaN  1/24/1980        12:13 PM  113590    3.055   

   Senior Management                  Team  
22              True       Client Services  
27             False                 Legal  
31             False               Product  
41              True  Business Development  
49             False                 Sales

As we can see above. 20th row is not recognized as a NaN value. Pandas recognizes the “NA” as a missing value. Unfortunately, the other types weren’t recognized. If there’s a couple of users entering data or you combine data from different sources, 
then this is a common problem. 

A way to solve that problem is using a list:

In [12]:
# Making a list of missing value types
missing_values = ["n/a", "na", "--"]
data= pd.read_csv("employees.csv", na_values = missing_values)

# creating bool series True for NaN values  
bool_series = pd.isnull(data["Gender"]) 

# displaying data only with Gender = NaN  
data[bool_series].head()

First Name Gender Start Date Last Login Time  Salary  Bonus %  \
20       Lois    NaN  4/22/1995         7:18 PM   64714    4.934   
22     Joshua    NaN   3/8/2012         1:58 AM   90816   18.816   
27      Scott    NaN  7/11/1991         6:58 PM  122367    5.218   
31      Joyce    NaN  2/20/2005         2:40 PM   88657   12.752   
41  Christine    NaN  6/28/2015         1:08 AM   66582   11.308   

   Senior Management                  Team  
20              True                 Legal  
22              True       Client Services  
27             False                 Legal  
31             False               Product  
41              True  Business Development

As you can see the 20 th row is changed again. 

To handle these non-standard types of missing data is crucial before you want to perform operations on missing data. Otherwise 
you can have suprises along the way.

It is good to go through the data and make a comprehensive list for all types of missing data.

 # Checking for missing values using notnull()

In order to check null values in Pandas Dataframe, we use notnull() function this function return dataframe of Boolean values which are False for NaN values.

In [3]:
# creating bool series True for NaN values  
bool_series = pd.notnull(data["Gender"])  
    
# displayind data only with Gender = Not NaN  
data[bool_series].head() 

First Name  Gender Start Date Last Login Time  Salary  Bonus %  \
0    Douglas    Male   8/6/1993        12:42 PM   97308    6.945   
1     Thomas    Male  3/31/1996         6:53 AM   61933    4.170   
2      Maria  Female  4/23/1993        11:17 AM  130590   11.858   
3      Jerry    Male   3/4/2005         1:00 PM  138705    9.340   
4      Larry    Male  1/24/1998         4:47 PM  101004    1.389   

  Senior Management             Team  
0              True        Marketing  
1              True              NaN  
2             False          Finance  
3              True          Finance  
4              True  Client Services

In [24]:
data[data.notnull()].head()

First Name  Gender Start Date Last Login Time  Salary  Bonus %  \
0    Douglas    Male   8/6/1993        12:42 PM   97308    6.945   
1     Thomas    Male  3/31/1996         6:53 AM   61933    4.170   
2      Maria  Female  4/23/1993        11:17 AM  130590   11.858   
3      Jerry    Male   3/4/2005         1:00 PM  138705    9.340   
4      Larry    Male  1/24/1998         4:47 PM  101004    1.389   

  Senior Management             Team  
0              True        Marketing  
1              True              NaN  
2             False          Finance  
3              True          Finance  
4              True  Client Services

In [48]:
#Select the rows of data where gender is not NaN and team is not NaN

data[data['Gender'].notnull() & data['Team'].notnull()].head()

First Name  Gender Start Date Last Login Time  Salary  Bonus %  \
0    Douglas    Male   8/6/1993        12:42 PM   97308    6.945   
2      Maria  Female  4/23/1993        11:17 AM  130590   11.858   
3      Jerry    Male   3/4/2005         1:00 PM  138705    9.340   
4      Larry    Male  1/24/1998         4:47 PM  101004    1.389   
5     Dennis    Male  4/18/1987         1:35 AM  115163   10.125   

  Senior Management             Team  
0              True        Marketing  
2             False          Finance  
3              True          Finance  
4              True  Client Services  
5             False            Legal

# Summarizing Missing Values

If you want to look at the total number of missing values for each feature, you can do the following:

In [21]:
# making data frame from csv file  
data = pd.read_csv("employees.csv") 

# Total missing values for each feature
print (data.isnull().sum())

First Name            67
Gender               145
Start Date             0
Last Login Time        0
Salary                 0
Bonus %                0
Senior Management     67
Team                  43
dtype: int64


In [22]:
#Other times we might want to do a quick check to see if we have any missing values at all.

print (data.isnull().values.any())

True


In [23]:
# We might also want to get a total count of missing values.

print (data.isnull().sum().sum())

322


# Dropping missing values using dropna()

In order to drop a null values from a dataframe, we used dropna() function.

This function drop Rows/Columns of datasets with Null values in different ways.

We cannot drop single values from a DataFrame; we can only drop full rows or full columns. Depending on the application, you might want one or the other, so dropna() gives a number of options for a DataFrame.

By default, dropna() will drop all rows in which any null value is present:

Dropping rows with at least 1 null value.

In [13]:
# making new data frame with dropped NA values  
new_data = data.dropna(axis = 0, how ='any')  
    
new_data.head()

First Name  Gender Start Date Last Login Time  Salary  Bonus %  \
0    Douglas    Male   8/6/1993        12:42 PM   97308    6.945   
2      Maria  Female  4/23/1993        11:17 AM  130590   11.858   
3      Jerry    Male   3/4/2005         1:00 PM  138705    9.340   
4      Larry    Male  1/24/1998         4:47 PM  101004    1.389   
5     Dennis    Male  4/18/1987         1:35 AM  115163   10.125   

  Senior Management             Team  
0              True        Marketing  
2             False          Finance  
3              True          Finance  
4              True  Client Services  
5             False            Legal

In [14]:
# Now we compare sizes of data frames so that we can come to know how many rows had at least 1 Null value

print("Old data frame length:", len(data)) 
print("New data frame length:", len(new_data))  
print("Number of rows with at least 1 NA value: ", (len(data)-len(new_data))) 


Old data frame length: 1000
New data frame length: 764
Number of rows with at least 1 NA value:  236


Since the difference is 236, there were 236 rows which had at least 1 Null value in any column.


Alternatively, you can drop NA values along a different axis; axis=1 drops all columns containing a null value.


But this drops some good data as well; you might rather be interested in dropping rows or columns with all NA values, or a majority of NA values. This can be specified through the how or thresh parameters, which allow fine control of the number of nulls to allow through.

The default is how='any', such that any row or column (depending on the axis keyword) containing a null value will be dropped. You can also specify how='all', which will only drop rows/columns that are all null values.

To keep only the rows with at least 2 non-NA values, you need to use 'tresh' option: df.dropna (thresh=2)


In [16]:
new_data_all= data.dropna(axis=1, how='all')
new_data_all.head()

First Name  Gender Start Date Last Login Time  Salary  Bonus %  \
0    Douglas    Male   8/6/1993        12:42 PM   97308    6.945   
1     Thomas    Male  3/31/1996         6:53 AM   61933    4.170   
2      Maria  Female  4/23/1993        11:17 AM  130590   11.858   
3      Jerry    Male   3/4/2005         1:00 PM  138705    9.340   
4      Larry    Male  1/24/1998         4:47 PM  101004    1.389   

  Senior Management             Team  
0              True        Marketing  
1              True              NaN  
2             False          Finance  
3              True          Finance  
4              True  Client Services

In [17]:
new_data_all= data.dropna(axis=1, how='any')
new_data_all.head()

Start Date Last Login Time  Salary  Bonus %
0   8/6/1993        12:42 PM   97308    6.945
1  3/31/1996         6:53 AM   61933    4.170
2  4/23/1993        11:17 AM  130590   11.858
3   3/4/2005         1:00 PM  138705    9.340
4  1/24/1998         4:47 PM  101004    1.389

In [24]:
new_data_all= data.dropna(axis=1,thresh=2)
new_data_all.head()

First Name  Gender Start Date Last Login Time  Salary  Bonus %  \
0    Douglas    Male   8/6/1993        12:42 PM   97308    6.945   
1     Thomas    Male  3/31/1996         6:53 AM   61933    4.170   
2      Maria  Female  4/23/1993        11:17 AM  130590   11.858   
3      Jerry    Male   3/4/2005         1:00 PM  138705    9.340   
4      Larry    Male  1/24/1998         4:47 PM  101004    1.389   

  Senior Management             Team  
0              True        Marketing  
1              True              NaN  
2             False          Finance  
3              True          Finance  
4              True  Client Services

# Filling missing values using fillna()

In order to fill null values in a datasets, we use fillna(), replace() and interpolate() function these function replace NaN values with some value of their own.

The idea behind filling data with some value must be supported by a statistical reasoning. So according to your approach,
this value might be a single number like zero, or it might be some sort of imputation or interpolation from the good values. 

In [55]:
# Now we are going to fill all the null values in Gender column with “No Gender”
# filling a null values using fillna()  
data["Gender"].fillna("No Gender", inplace = True)  
  
data[20:24]

First Name     Gender Start Date Last Login Time  Salary  Bonus %  \
20       Lois  No Gender  4/22/1995         7:18 PM   64714    4.934   
21    Matthew       Male   9/5/1995         2:12 AM  100612   13.645   
22     Joshua  No Gender   3/8/2012         1:58 AM   90816   18.816   
23        NaN       Male  6/14/2012         4:19 PM  125792    5.042   

   Senior Management             Team  
20              True            Legal  
21             False        Marketing  
22              True  Client Services  
23               NaN              NaN

In [56]:
#We can fill NA entries with a single value, such as zero:

import numpy as np

data2 = pd.Series([1, np.nan, 2, None, 3, 8, None], index=list('ABCDEFG'))
print(data2)

print(data2.fillna(0))

# Or you can use the mean or median which are commonly used

print (data2.fillna(data2.mean()))

A    1.0
B    NaN
C    2.0
D    NaN
E    3.0
F    8.0
G    NaN
dtype: float64
A    1.0
B    0.0
C    2.0
D    0.0
E    3.0
F    8.0
G    0.0
dtype: float64
A    1.0
B    3.5
C    2.0
D    3.5
E    3.0
F    8.0
G    3.5
dtype: float64


In [57]:
#We can specify a forward-fill to propagate the previous value forward:

data2.fillna(method='ffill')

A    1.0
B    1.0
C    2.0
D    2.0
E    3.0
F    8.0
G    8.0
dtype: float64

In [58]:
#Or we can specify a back-fill to propagate the next values backward:

data2.fillna(method='bfill')

A    1.0
B    2.0
C    2.0
D    3.0
E    3.0
F    8.0
G    NaN
dtype: float64

For DataFrames, the options are similar, but we can also specify an axis along which the fills take place. Be aware that if a previous value is not available during a forward fill, the NA value remains.

# Filling null values using replace() method

    

In [59]:
# Printing the first 10 to 24 rows of 
# the data frame for visualization    
data[23:25] 

First Name Gender Start Date Last Login Time  Salary  Bonus %  \
23        NaN   Male  6/14/2012         4:19 PM  125792    5.042   
24       John   Male   7/1/1992        10:08 PM   97950   13.873   

   Senior Management             Team  
23               NaN              NaN  
24             False  Client Services

In [60]:
# Now we are going to replace the all Nan value in the data frame with -99 value.

# will replace  Nan value in dataframe with value -99   

data.replace(to_replace = np.nan, value = -99)[23:25]


First Name Gender Start Date Last Login Time  Salary  Bonus %  \
23        -99   Male  6/14/2012         4:19 PM  125792    5.042   
24       John   Male   7/1/1992        10:08 PM   97950   13.873   

   Senior Management             Team  
23               -99              -99  
24             False  Client Services

In [44]:
# Replace one value to another
# Need to use inplace=True to mutate the dataframe itself

df = pd.DataFrame({
    'name':['ali','mehmet','yusuf'],
    'age':[25,35,68],
    'city':['istanbul','ankara','izmir']
})

df.replace([68],86)


name  age      city
0     ali   25  istanbul
1  mehmet   35    ankara
2   yusuf   86     izmir

In [45]:
# Replace with dictionary

df.replace({
    25:26,
    'ali':'aliye'
})


name  age      city
0   aliye   26  istanbul
1  mehmet   35    ankara
2   yusuf   68     izmir

In [46]:
# Replace with regex

df.replace('al.+','XYZ',regex=True)

name  age      city
0     XYZ   25  istanbul
1  mehmet   35    ankara
2   yusuf   68     izmir

In [47]:
#If we want to replace more than one value such as 'A' or 'B' then we can also use the list:
df.replace(to_replace =["istanbul", "yusuf"],value ="XXXX") 

name  age    city
0     ali   25    XXXX
1  mehmet   35  ankara
2    XXXX   68   izmir

# Using interpolate() function to fill the missing values

There are different alternatives to use interpolate function. As discussed earlier, you should use statistically 
robust methods and substantiate your reasoning. Below are some examples of linear method.

In [51]:
# Creating the dataframe   
df2= pd.DataFrame({"A":[45, 47, 59, None, 18, 57],  
                   "B":[None, 26, 54, 39, None, 67],  
                   "C":[76, 56, None, 32, 81, 56],  
                   "D":[140, 39, None, None, 65, 87]})  
    
# Print the dataframe  
df2  


A     B     C      D
0  45.0   NaN  76.0  140.0
1  47.0  26.0  56.0   39.0
2  59.0  54.0   NaN    NaN
3   NaN  39.0  32.0    NaN
4  18.0   NaN  81.0   65.0
5  57.0  67.0  56.0   87.0

In [52]:
# Let’s interpolate the missing values using Linear method. 
# Note that Linear method ignore the index and treat the values as equally spaced.

# to interpolate the missing values  
df2.interpolate(method ='linear', limit_direction ='forward') 


A     B     C           D
0  45.0   NaN  76.0  140.000000
1  47.0  26.0  56.0   39.000000
2  59.0  54.0  44.0   47.666667
3  38.5  39.0  32.0   56.333333
4  18.0  53.0  81.0   65.000000
5  57.0  67.0  56.0   87.000000

As we can see the output, values in the first row could not get filled as the direction of filling of values is forward and there is no previous value which could have been used in interpolation.

Use interpolate() function to interpolate the missing values in the backward direction using linear method and putting a limit on maximum number of consecutive Na values that could be filled.

In [54]:
# to interpolate the missing values 

df2.interpolate(method ='linear', limit_direction ='backward', limit = 1) 

A     B     C           D
0  45.0  26.0  76.0  140.000000
1  47.0  26.0  56.0   39.000000
2  59.0  54.0  44.0         NaN
3  38.5  39.0  32.0   56.333333
4  18.0  53.0  81.0   65.000000
5  57.0  67.0  56.0   87.000000

Notice the third column, only one missing value has been filled as we have put the limit to 1. 

# RESOURCES:

1-John Sullivan,Data Scientist | Pizza Lover | Bulldog Father | dataoptimal.com
(https://towardsdatascience.com/data-cleaning-with-python-and-pandas-detecting-missing-values-3e9c6ebcf78b)

2- https://www.geeksforgeeks.org/working-with-missing-data-in-pandas/

3- https://jakevdp.github.io/PythonDataScienceHandbook/03.04-missing-values.html

4- https://stackoverflow.com/questions/54618253/how-to-deal-with-missing-values-in-pandas-dataframe
